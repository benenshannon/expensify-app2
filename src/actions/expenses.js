import uuid from 'uuid';
import database from '../firebase/firebase';

//How the action generators work:
//component calls action generator
//action generator returns an object
//component dispatches the object
//redux store changes

//ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        } = expenseData;
        const expense = {description, note, amount, createdAt};

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

//REMOVE_EXPENSE
export const removeExpense = (
    { id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//startRemoveExpense
export const startRemoveExpense= ({ id } = {}) => {
    return (dispatch) => {
        //dispatch gets passed to this function by the redux library
        return database.ref(`expenses/${id}`).remove().then(() => {
            //once the item has been removed, the removeExpense action
            //generator is then called
            dispatch(removeExpense( { id }));
        });
    };
};

//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// manipulate the redux store
//SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

//THIS WILL ACTUALLY FETCH THE DATA
//export const startSetExpenses;
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];

            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setExpenses(expenses));
        });
    };
};