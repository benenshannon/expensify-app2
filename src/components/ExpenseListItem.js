import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

//removeExpense already has the function built in


const ExpenseListItem = ({id, description, amount, createdAt}) => (
    <div>
      <Link to={`/edit/${id}`}>
         <h3>{description}</h3> 
      </Link>
      <p>
        {numeral(amount / 100 ).format('$0,0.00')}
        -
        {moment(createdAt).format('Do MMM, YYYY')}
      </p>
    </div>
    //remember, when calling a function (removeExpense) to pass in the relevant
    // argument - see the removeExpense code in actions/expenses.js, where
    //id is required item
);

//always include "connect()" as this connects to the connected component
export default ExpenseListItem;

//This is the longer way of doing it
//const ExpenseListItem = (props) => (
//    <div>
//      {props.expenses.map((expense) => 
//        <li key = {expense.id}>
//        <h3 >Description: {expense.description}</h3>
//        <p>Amount: {expense.amount}</p>
//        <p>CreatedAt: {expense.createdAt}</p>
//        </li>)};
//    </div>
//);

//mapping the state (redux store) to the props (in longer format than the 
//one in the "ExpenseList.js" file)
//in the connect() brackets we define a function
//that identifies the info we want to call
//const ConnectedExpenseListItem = connect((state) => {
//    return {
//        expenses: state.expenses
//    };
//})(ExpenseListItem);

//export default ConnectedExpenseListItem;