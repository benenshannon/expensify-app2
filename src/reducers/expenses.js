// Expenses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState , action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            //.filter is a way of searching an array for
            //objects that meet a criteria - such as !== action.id
            // and returning the new filtered array 
            return state.filter(( { id }) => id !== action.id);
            //return state.filter(( { id }) => {return id !== action.id});
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id===action.id){
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                };
            });
        case 'SET_EXPENSES':
        // returning the action object with the expense array attached
            return action.expenses;   
        default:
            return state;
    }
};