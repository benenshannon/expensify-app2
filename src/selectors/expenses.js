import moment from 'moment';

//Get visible expenses
export default (expenses, {text, sortBy, startDate, endDate }) => {
    //arrow function calling "expenses" array, and 
    //"filters" (destructured by using {})
        return expenses.filter((expense) => {
            const createdAtMoment = moment(expense.createdAt)
            const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true ;
           const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true ;
            const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
            return startDateMatch && endDateMatch && textMatch;
        }).sort((a, b) => {
            if(sortBy === 'date') {
                return a.createdAt < b.createdAt ? 1 : -1;
                //if a is less than than b, then true = 1, meaning b will come first
            }
            else if(sortBy === 'amount') {
                return a.amount < b.amount ? 1 : -1;
            }
        });
    }; 
    