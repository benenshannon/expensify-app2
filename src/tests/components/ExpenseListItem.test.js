import React from 'react';
import { shallow } from 'enzyme';
import "../setupTests";
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';
 
test('should render expense list item with expenses', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    //ExpenseListItem expects several props - id, desc, amount, createdAt
    //so using the "..." spreads out the expenses object, and adds each one
    //as a prop to ExpenseListItem
    expect(wrapper).toMatchSnapshot();
});
