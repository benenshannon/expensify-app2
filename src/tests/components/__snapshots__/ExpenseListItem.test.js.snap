// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list item with expenses 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      rent
    </h3>
  </Link>
  <p>
    189500
    -
    1st Jan, 1970
  </p>
</div>
`;
